File: LeetCode/binaryTree/652.find-duplicate-subtrees.py
#
# @lc app=leetcode id=652 lang=python3
#
# [652] Find Duplicate Subtrees
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    uniqueStructure = {}
    dupStrucutre = []

    def serialization(self, root: Optional[TreeNode]) -> string:
        if root is None: return '#,'
        leftSerial = self.serialization(root.left)
        rightSerial = self.serialization(root.right)

        # got unique structure ID
        mySerial = leftSerial + rightSerial + str(root.val) + ','

        # compare mySerial with global record
        if(mySerial in self.uniqueStructure): # only append to res in first reappear
            if self.uniqueStructure[mySerial] == 1:
                self.uniqueStructure[mySerial]+=1
                self.dupStrucutre.append(root)
        else: # first time: create record
            self.uniqueStructure[mySerial] = 1

        return mySerial

    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        # re-init global vars
        self.uniqueStructure = {}
        self.dupStrucutre = []
        self.serialization(root)
        print(self.uniqueStructure)

        return self.dupStrucutre




# @lc code=end

# 序列化遍历整个树得到以任意一个node为树根的子树的结构, 序列化 :=> 任何相同的结构应该都有相同的序列
# 全局变量uniqueStructure (hash map)存储出现过的，unique的结构（序列），
# 如果任意一个node的子树序列在uniqueStructure出现过一次，就在结果中加上她

# 后序遍历！！ 这样保证获得整个子树结构后，在树根进行判断

File: LeetCode_Go/BinaryTree/652.find-duplicate-subtrees.go
/*
 * @lc app=leetcode id=652 lang=golang
 *
 * [652] Find Duplicate Subtrees
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Given a binary tree, return all duplicate subtrees in an array. For each kind of duplicate subtrees, return the root node of any one of the them. Two trees are duplicate if they have the same structure with the same node values.

// Serialize tree with postorder dfs + null string, determine identicality by comparing the serialized strings. Use a hashmap to store serialized strings and determine duplication.

// O(n) time | O(n) space

var SEP string = ","
var NULL string = "#"

func findDuplicateSubtrees(root *TreeNode) []*TreeNode {
	var sequenceFreqMap map[string]int = make(map[string]int)
	var duplicateSubtrees []*TreeNode = make([]*TreeNode, 0)

	_ = treeSerializerHelper(root, sequenceFreqMap, &duplicateSubtrees)

	// for key, value := range sequenceFreqMap {
	// 	fmt.Printf("key: %s, value: %d\n", key, value)
	// }

	return duplicateSubtrees
}

func treeSerializerHelper(root *TreeNode, sequenceFreqMap map[string]int, duplicateSubtrees *[]*TreeNode) string {
	if root == nil {
		return NULL + SEP
	}

	leftSeq := treeSerializerHelper(root.Left, sequenceFreqMap, duplicateSubtrees)
	rightSeq := treeSerializerHelper(root.Right, sequenceFreqMap, duplicateSubtrees)

	rootSeq := leftSeq + rightSeq + strconv.Itoa(root.Val) + SEP

	if _, exists := sequenceFreqMap[rootSeq]; exists {
		sequenceFreqMap[rootSeq] += 1
		if sequenceFreqMap[rootSeq] == 2 { // first time a duplication found
			*duplicateSubtrees = append(*duplicateSubtrees, root)
		}
	} else {
		sequenceFreqMap[rootSeq] = 1
	}

	return rootSeq
}

// @lc code=end

